federation-js/src/joinSpec.ts

```diff
@@ -0,0 +1,83 @@
+import {
+  GraphQLDirective,
+  DirectiveLocation,
+  GraphQLEnumType,
+  GraphQLScalarType,
+} from 'graphql';
+import { ServiceDefinition } from './composition';
+
+const FieldSetScalar = new GraphQLScalarType({
+  name: 'join__FieldSet',
+});
+
+function getJoinGraphEnum(serviceList: ServiceDefinition[]) {
+  return new GraphQLEnumType({
+    name: 'join__Graph',
+    values: Object.fromEntries(
+      serviceList.map((service) => [
+        service.name.toUpperCase(),
+        { value: service.url },
+      ]),
+    ),
+  });
+}
+
+function getJoinFieldDirective(JoinGraphEnum: GraphQLEnumType) {
+  return new GraphQLDirective({
+    name: 'join__field',
+    locations: [DirectiveLocation.FIELD_DEFINITION],
+    args: {
+      graph: {
+        type: JoinGraphEnum,
+      },
+      requires: {
+        type: FieldSetScalar,
+      },
+      provides: {
+        type: FieldSetScalar,
+      },
+    },
+  });
+}
+
+function getJoinOwnerDirective(JoinGraphEnum: GraphQLEnumType) {
+  return new GraphQLDirective({
+    name: 'join__owner',
+    locations: [DirectiveLocation.OBJECT, DirectiveLocation.INTERFACE],
+    args: {
+      graph: {
+        type: JoinGraphEnum,
+      },
+    },
+  });
+}
+
+export function getJoins(serviceList: ServiceDefinition[]) {
+  const JoinGraphEnum = getJoinGraphEnum(serviceList);
+  const JoinFieldDirective = getJoinFieldDirective(JoinGraphEnum);
+  const JoinOwnerDirective = getJoinOwnerDirective(JoinGraphEnum);
+
+  const JoinTypeDirective = new GraphQLDirective({
+    name: 'join__type',
+    locations: [DirectiveLocation.OBJECT, DirectiveLocation.INTERFACE],
+    args: {
+      graph: {
+        type: JoinGraphEnum,
+      },
+      requires: {
+        type: FieldSetScalar,
+      },
+      provides: {
+        type: FieldSetScalar,
+      },
```

Comment:
The argument needs to be `key`, and the `@join__type` directive should be repeatable (I fixed this on the `typescript-redux` branch, so maybe we could cherry pick a09e4339b6232ce400a40edc9e0694b681f4b7a2).

Reply:
Good catch, `cherry-pick`ed a09e433!